### PR TITLE
New module for AWS CodeBuild

### PR DESCRIPTION
**Pull request by stefanhorning**
##### SUMMARY
New module aws_codebuild for the AWS CodeBuild service.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
aws_codebuild.yml module

### CODE REVIEW

**Comment by stefanhorning** [comment]
Sucessor of PR https://github.com/ansible/ansible/pull/38943

**Comment by stefanhorning** [comment]
Gist with testplay https://gist.github.com/stefanhorning/d168eb3b4d13e828a9fd4b37d5dd3379

**Comment by stefanhorning** [comment]
Ready to merge from my side (I an do a rebase once reviewed, if there is a wish)

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]:
```suggestion
    pass  # Handled by AnsibleAWSModule
```

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]: This looks suspicious; I don't see any source for an exception, so you shouldn't be trying to pass one to fail_json. This check is also unnecessary, since the 'name' argument is marked as required in the argspec.

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (flowerysong) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]: This (and all of the other exception handling) should be using module.fail_json_aws() instead of doing manual exception formatting.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (flowerysong) [inline test/integration/targets/aws_codebuild/tasks/main.yml:19]: This role should probably be cleaned up after the tests are done. (Also, I don't have any experience specifically with CodeBuild, but based on general experience with AWS you may need to insert a short pause after the IAM role is created to avoid sporadic failures when trying to use a newly created role.)

**Review by flowerysong — CHANGES_REQUESTED** [review]

**Comment by stefanhorning** [comment]
@flowerysong applied all your recommended changes now.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (ptux) [inline test/integration/targets/aws_codebuild/tasks/main.yml]: All cleanup tasks should be done in `always` block to avoid that cleanup does not execute when failed.

**Review by ptux — CHANGES_REQUESTED** [review]
All cleanup tasks should be done in always block to avoid that cleanup does not execute when failed.

File: test/integration/targets/aws_codebuild/aliases
Comment (ptux) [inline test/integration/targets/aws_codebuild/aliases]: You are turning the integration tests off.
So, the integration tests you wrote are not executed.

**Review by ptux — CHANGES_REQUESTED** [review]

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [inline test/integration/targets/aws_codebuild/tasks/main.yml]: The CI testrunner fails to create this IAM role (AccessDenied), so what do I need to do to get this working?

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_codebuild/tasks/main.yml:19]: Done

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (mattclay) [reply test/integration/targets/aws_codebuild/tasks/main.yml]: We'll need to update our CI system to accommodate the additional permissions required.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/aws_codebuild/tasks/main.yml]: This is already provided in the credentials so it causes a warning for the duplicate key.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/aws_codebuild/tasks/main.yml]: Another duplicate region parameter

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]: I'd like to have a whitelist of expected parameters, rather than popping off those that are unexpected. This is buggy right now since you can pass in unsupported params (such as aws_access_key). If other parameters are ever added to the module or doc fragments this could continue causing possible issues. If you don't want to hardcode your whitelist you can make it dynamic:
```
from ansible.module_utils.aws.core import get_boto3_client_method_parameters

formatted_params = snake_dict_to_camel_dict(dict((k, v) for k, v in params.items() if v is not None))

permitted_update_params = get_boto3_client_method_parameters(client, 'update_project')
permitted_create_params = get_boto3_client_method_parameters(client, 'create_project')

formatted_clean_update_params = dict((k, v) for k, v in formatted_params.items() if k in permitted_update_params)
formatted_clean_create_params = dict((k, v) for k, v in formatted_params.items() if k in permitted_create_params)
```

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/aws_codebuild/tasks/main.yml]: This delay is too short so the tests are repeatedly failing for me

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/aws_codebuild/tasks/main.yml:109]: `source` and `artifacts` are required by the module so this task fails

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (s-hertel) [inline test/integration/targets/aws_codebuild/tasks/main.yml:83]: aws_account_id is undefined

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (stefanhorning) [reply lib/ansible/modules/cloud/amazon/aws_codebuild.py]: Yes, that was still my first hacky solution, thanks for catching that. Also didn't know there was a convinient way already to determine valid param. Updated now

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_codebuild/tasks/main.yml]: Increased now. Hovewer the CI runner is not yet updated, so I can't iron out any other issues with the tests at this point, but will certainly do so if we get the IAM role creation approved.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_codebuild/tasks/main.yml:83]: This should also be fixed now

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_codebuild/tasks/main.yml:109]: Fixed all your comments except this one for now. Will wait until the CI runner is updated (to create the IAM role), so I can iron out any remaining issues with the tests.

**Comment by stefanhorning** [comment]
ready_for_review

**Comment by stefanhorning** [comment]
Any chance to get the CI update done? Otherwise I will just disable the tests for now to get this green and merged.

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [reply test/integration/targets/aws_codebuild/tasks/main.yml:109]: Also updated now

**Comment by stefanhorning** [comment]
@s-hertel I responded to the remaining comment of you now and also turned the build green. This can be merged now.

**Comment by kapilt** [comment]
so the primary change needed is just adding `ansible-test-sts-` before the {resource_prefix}

**Comment by stefanhorning** [comment]
@kapilt thanks for looking into this. I enabled tests again with you recommended chang. However there still seems to be an AWS permissions issue when creating the IAM role:

```
User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=122139.100 is not authorized to perform: iam:CreateInstanceProfile on resource: arn:aws:iam::966509639900:instance-profile/ansible-test-sts-shippable-122139-100-codebuild-service-role
```

File: test/integration/targets/aws_codebuild/tasks/main.yml
Comment (stefanhorning) [inline test/integration/targets/aws_codebuild/tasks/main.yml:23]: I guess this happens when trying to attach the `assume_role_policy_document` here, which I think we need for the aws_code_build module to accept this role.

**Comment by stefanhorning** [comment]
@kapilt I think the issue here is that according to hacking/aws_config/testing_policies/sts-policy.json IAM users with a prefix `ansible-test-sts-` are allowed to in fact to a iam:CreateInstanceProfile action, but from the error message here it seems that action is actually be done by arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=122139.100 which I am not sure has the same permissions.

**Comment by mattclay** [comment]
@stefanhorning It looks like the permissions issues have been resolved. Tests are now failing due to an undefined variable: https://app.shippable.com/github/ansible/ansible/runs/125395/100/tests

**Comment by stefanhorning** [comment]
Thanks @mattclay and @kapilt for finally sorting out build ENV issues. I was now finally able to turn all tests green including integration tests! :tada: 

Should I rebase this branch once more to cleanup commits? Or can we just merge it now?

**Comment by mattclay** [comment]
@stefanhorning No need to squash commits. That will be done when the PR is merged.

**Comment by stefanhorning** [comment]
Can we finally merge this PR then?

**Comment by mattclay** [comment]
@s-hertel Have all of your review comments been addressed?

**Comment by stefanhorning** [comment]
@mattclay Yes, I addressed all @s-hertel s comments already, see https://github.com/ansible/ansible/pull/47187#pullrequestreview-174103001

Already asked him a month ago to lift his 'requested changes' status on this PR, but he didn't respond since...

**Comment by s-hertel** [comment]
@stefanhorning I'm sorry. I've been working on AWS related things very minimally and the Github notifications fell through the cracks. Pinging me on IRC tends to be a good way to get my immediate attention.

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]:
```suggestion

```

File: lib/ansible/modules/cloud/amazon/aws_codebuild.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_codebuild.py]:
```suggestion

```

**Review by s-hertel — APPROVED** [review]

**Comment by stefanhorning** [comment]
@s-hertel no worries, it's an opensource project after all ;)

Commited your requested changes now, I guess we can merge this once the build is green.

**Comment by s-hertel** [comment]
Thanks for sticking with it, and everyone involved for reviews and CI help. :-)

**Pull request merged**
